app/code/Magento/Catalog/Model/Product/Image/Process/Queue.php

```diff
@@ -0,0 +1,322 @@
+        ProductRepositoryInterface $productRepository,
+        ResourceConnection $resourceConnection,
+        CacheFactory $imageCacheFactory,
+        OutputInterface $output,
+        $maxProcesses = self::DEFAULT_MAX_PROCESSES_AMOUNT
+    ) {
+        $this->productRepository = $productRepository;
+        $this->resourceConnection = $resourceConnection;
+        $this->output = $output;
+        $this->imageCacheFactory = $imageCacheFactory;
+        $this->maxProcesses = $maxProcesses;
+        $this->imageCache = $this->imageCacheFactory->create();
+    }
+
+    /**
+     * @param array $productIds
+     * @return Queue
+     */
+    public function setProducts(array $productIds)
+    {
+        $this->products = array_combine($productIds, $productIds);
+        return $this;
+    }
+
+    /**
+     * @return Product[]
+     */
+    public function getProducts()
+    {
+        return $this->products;
+    }
+
+    /**
+     * Process jobs
+     *
+     * @return int
+     */
+    public function process()
+    {
+        if ($this->isCanBeParalleled()) {
+            $this->imageCache->preloadData();
+        }
+        $returnStatus = 0;
+        $this->start = $this->lastJobStarted = time();
+        $products = $this->products;
+        while (count($products)) {
+            foreach ($products as $productId) {
+                $this->assertAndExecute($productId, $products);
+            }
+            usleep(20000);
+            foreach ($this->inProgress as $productId) {
+                if ($this->isResized($productId)) {
+                    unset($this->inProgress[$productId]);
+                }
+            }
+        }
+
+        $this->awaitForAllProcesses();
+
+        return $returnStatus;
+    }
+
+    /**
+     * Check that we are not over max processes and execute
+     *
+     * @param int $productId
+     * @param array $products
+     * @return void
+     */
+    private function assertAndExecute($productId, array & $products)
+    {
+        if ($this->maxProcesses < 2 || (count($this->inProgress) < $this->maxProcesses)) {
+            unset($products[$productId]);
+            $this->execute($productId);
+        }
+    }
+
+    /**
+     * Wait untill all processes are finished
+     *
+     * @return void
+     */
+    private function awaitForAllProcesses()
+    {
+        while ($this->inProgress) {
+            foreach ($this->inProgress as $productId => $product) {
+                if ($this->isResized($product)) {
+                    unset($this->inProgress[$productId]);
+                }
+            }
+            usleep(100000);
+        }
+        if ($this->isCanBeParalleled()) {
+            // close connections only if ran with forks
+            $this->resourceConnection->closeConnection();
+        }
+    }
+
+    /**
+     * @return bool
+     */
+    private function isCanBeParalleled()
+    {
+        return function_exists('pcntl_fork') && $this->maxProcesses > 1;
```

Comment:
Please look into usage of `popen`, if possible, as we have encountered issues with parallelization with `pcntl_fork` in the past (for instance Redis session management may not really work as expected in this case).